docs/fsharp/language-reference/byrefs.md

```diff
@@ -0,0 +1,243 @@
+
+|C# construct|F# infers|
+|------------|---------|
+|`ref` return value|`outref<'T>`|
+|`ref readonly` return value|`inref<'T>`|
+|`in ref` parameter|`inref<'T>`|
+|`out ref` parameter|`outref<'T>`|
+
+The following table shows what F# emits:
+
+|F# construct|Emitted construct|
+|------------|-----------------|
+|`inref<'T>` argument|`[In]` attribute on argument|
+|`inref<'T>` return|`modreq` attribute on value|
+|`inref<'T>` in abstract slot or implementation|`modreq` on argument or return|
+|`outref<'T>` argument|`[Out]` attribute on argument|
+
+### Type inference and overloading rules
+
+An `inref<'T>` type is inferred by the F# compiler in the following cases:
+
+1. A .NET parameter or return type that has an `IsReadOnly` attribute.
+2. The `this` pointer on a struct type that has no mutable fields.
+3. The address of a memory location derived from another `inref<_>` pointer.
+
+When an implicit address of an `inref` is being taken, an overload with an argument of type `SomeType` is preferred to an overload with an argument of type `inref<SomeType>`. For example:
+
+```fsharp
+type C() =
+    static member M(x: System.DateTime) = x.AddDays(1.0)
+    static member M(x: inref<System.DateTime>) = x.AddDays(2.0)
+    static member M2(x: System.DateTime, y: int) = x.AddDays(1.0)
+    static member M2(x: inref<System.DateTime>, y: int) = x.AddDays(2.0)
+
+let res = System.DateTime.Now
+let v =  C.M(res)
+let v2 =  C.M2(res, 4)
+```
+
+In both cases, the overloads taking `System.DateTime` are resolved rater than the overloads taking `inref<System.DateTime>`.
+
+## Byref-like structs
+
+In addition to the `byref`/`inref`/`outref` trio, you can define your own structs that can adhere to `byref`-like semantics. This is done with the <xref:System.Runtime.CompilerServices.IsByRefLikeAttribute> attribute:
+
+```fsharp
+open System
+open System.Runtime.CompilerServices
+
+[<IsByRefLike; Struct>]
+type S(count1: Span<int>, count2: Span<int>) =
+    member x.Count1 = count1
+    member x.Count2 = count2
+```
+
+`IsByRefLike` does not imply `Struct`. Both must be present on the type.
+
+A "`byref`-like" struct in F# is a stack-bound value type. It is never allocated on the managed heap. They are useful for high-performance programming, as they are enforced with set of strong checks about the lifetimes and non-capture. These rules are:
+
+* They can be used as function parameters, method parameters, local variables, method returns.
+* They cannot be static or instance members of a class or normal struct.
+* They cannot be captured by any closure construct (`async` methods or lambda expressions).
+* They cannot be used as a generic parameter.
+
+This last point is crucial for F# "pipeline" style programming, as `|>` is a generic function that parameterizes its input types. This restriction may be relaxed for `|>` in the future, as it is inline and does not make any calls to non-inlined generic functions in its body.
```

Comment:
"pipeline" style --> pipeline-style (I don't think that pipeline needs to be in quotation marks)